keps/sig-node/4589-image-pull-source-support/README.md

```diff
@@ -0,0 +1,972 @@
+
+## Motivation
+
+As the Kubernetes is evolving in the cloud sector and users are willing to have their services offered in a disconnected/private environments, it is necessary to mirror the respective container images internally. Currently there is no easier way to determine the origin of the pulled image when image mirrors are configured other than going through the runtime logs.
+
+### Goals
+
+* Container Status section should show the actual pull source of the image
+* Inspecting Container Image using cri-tools should show the actual pull source of the image
+
+### Non-Goals
+
+* Not concerned about the image field in the container spec which may or may not be the same as the one being pulled by runtime.
+* Not concerned about the mirrors which are configured only for digest pulls.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+This KEP aims to introduce a new field in the Container Status as well as in the Container Image resections, inorder to provide the exact registry details of a image pulled by the runtime irrespective of the mirroring configuration. 
+* When no mirrors are configured, the registry details in the new field should be same as the one mentioned in the container spec section
+* When mirrors are configured, the image source details in the new field should be the one runtime used to pull that image.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+* When the imagePullPolicy is set to "Always" and when there are more than one 
+  mirrors configured for that image, the kubelet queries the container image registry 
+  to resolve the name to an image digest. If the kubelet has a container image with that exact 
+  digest cached locally, the kubelet uses its cached image and in this case, it is possible that 
+  the image digest might have been pulled from different source but the imagePullSource 
+  would still be pointing to an existing pulled source.
+* If the image is pulled outside of Kubernetes ecosystem, the imageSource field won't be set 
```

Comment:
arguably we would need to record and provide image source and redirects... if we are going to record a source we could also record import sources..

Reply:
This one is quite tricky for me to come up to some conclusion. I need expert ideas on this. If the image pull isn't under Kubernetes control, how can we handle it?